### ISSUE TITLE
PlacesHistoryStorageTest.kt test failures depending on locale

### ISSUE THREAD

**Issue by rfk** [mozilla/application-services#4209]
This is a test failure I found when running the android-components test suite, but I suspect it may be a bug in the underlying appservices layers somewhere.

The android-components `PlacesHistoryStorageTest.kt` suite has some tests that check page titles in non-English locales, such as [this one](https://github.com/mozilla-mobile/android-components/blob/d7b1b62da50cf224fce1eec7a163d858fb1a2334/components/browser/storage-sync/src/test/java/mozilla/components/browser/storage/sync/PlacesHistoryStorageTest.kt#L302). I have regularly observed intermittent test failures here when trying to smoketest a-c against a new a-s release, and finally decided to figure out why.

It appears to be due to my `$LANG` environment variable 😬 

If I run them like this, which is my default `$LANG` when using the native terminal on my Mac:

```
LANG="en_AU.UTF-8" ./gradlew browser-storage-sync:testDebugUnitTest
```

Then everything works fine and all tests pass. If I instead run them like this, which is my default `$LANG` when sshing in to my Mac from my Windows machine:

```
LANG="C.UTF-8" ./gradlew browser-storage-sync:testDebugUnitTest
```

Then this test fails:

```
  TEST: store can be used to track page meta information - title changes
    [Robolectric] mozilla.components.browser.storage.sync.PlacesHistoryStorageTest.store can be used to track page meta information - title changes: sdk=28; resources=binary
  FAILURE
org.junit.ComparisonFailure: expected:<[?????????]> but was:<[?????????]>
    at org.junit.Assert.assertEquals(Assert.java:115)
    at org.junit.Assert.assertEquals(Assert.java:144)
    at mozilla.components.browser.storage.sync.PlacesHistoryStorageTest$store can be used to track page meta information - title changes$1.invokeSuspend(PlacesHistoryStorageTest.kt:302)
```

I'm lightly alarmed that setting `$LANG` can affect the tests like this, and it makes me wonder if we may have some bug in the handoff of non-ascii strings between Kotlin and Rust.

Does this reproduce for anyone else?

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-2251)
┆Story Points: 3
┆Sprint: SYNC - 2021-06-25

**Comment by rfk** [comment]
I think we are messing up transfer of utf8 data from Kotlin to Rust via JNA. If I update our code to do an explicit `System.setProperty("jna.encoding", "UTF8")` and force JNA to treat all strings as UTF8 then the tests pass. So, I expect we're passing strings in the default system encoding, which is in turn depending on the value of `$LANG`.

**Comment by mhammond** [comment]
> I expect we're passing strings in the default system encoding,

Oof. I'm surprised this hasn't come up before

**Comment by rfk** [comment]
Me too :-/

Maybe there's some helpful default in Android that means this isn't actually biting us in the wild. But still, we should fix it for completeness.

**Comment by rfk** [comment]
> Maybe there's some helpful default in Android that means this isn't actually biting us in the wild

Aha, yep, according to the [official docs](https://developer.android.com/reference/java/nio/charset/Charset.html#defaultCharset()) the `Charset.defaultCharset()` method will always return `UTF-8` on Android devices. Phew.

**Comment by rfk** [comment]
Fixed by #4211.
